fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryWithSeparateContext.java

```diff
@@ -0,0 +1,185 @@
+
+/**
+ * An variant of {@link FDBDirectory} that has its own transactionality and can therefore operate on larger datasets for longer.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class FDBDirectoryWithSeparateContext extends FDBDirectory implements AutoCloseable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FDBDirectoryWithSeparateContext.class);
+
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+
+    private final Deque<CompletableFuture<?>> pendingReads = new ConcurrentLinkedDeque<>();
+
+    private long commitAfterBytesLimit;
+    private long commitAfterMillisLimit;
+    private long contextStartMillis;
+    private int pendingBytesCount;
+
+    @Nullable
+    IndexWriter indexWriter;
+
+    public FDBDirectoryWithSeparateContext(@Nonnull Subspace subspace, @Nonnull FDBRecordContext context) {
+        super(subspace, openLikeContext(context));
+        contextStartMillis = System.currentTimeMillis();
+    }
+
+    public void setCommitAfterLimits(long commitAfterBytesLimit, long commitAfterMillisLimit) {
+        this.commitAfterBytesLimit = commitAfterBytesLimit;
+        this.commitAfterMillisLimit = commitAfterMillisLimit;
+    }
+
+    @Override
+    public void close() {
+        if (indexWriter != null) {
+            try {
+                indexWriter.close();
+            } catch (IOException ex) {
+                throw new RecordCoreException("error closing writer", ex);
+            }
+        }
+        if (context != null) {
+            context.commit();
+        }
+        super.close();
+    }
+
+    @Override
+    protected <T> CompletableFuture<T> readTransaction(Function<ReadTransaction, CompletableFuture<T>> reader) {
+        Lock lock = readWriteLock.readLock();
+        lock.lock();
+        try {
+            CompletableFuture<T> result = super.readTransaction(reader);
+            pendingReads.removeIf(CompletableFuture::isDone);
+            if (!result.isDone()) {
+                pendingReads.addLast(result);
+            }
+            return result;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    protected void writeTransaction(Consumer<Transaction> writer) {
+        Lock lock = readWriteLock.writeLock();
+        lock.lock();
+        try {
+            super.writeTransaction(writer);
+            if ((commitAfterBytesLimit > 0 && pendingBytesCount > commitAfterBytesLimit) ||
+                    (commitAfterMillisLimit > 0 && System.currentTimeMillis() - contextStartMillis > commitAfterMillisLimit)) {
+                syncTransaction();
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    protected void countWriteBytes(long count) {
+        pendingBytesCount += count;
+    }
+
+    @Override
+    @SuppressWarnings("PMD.EmptyCatchBlock")
+    protected void syncTransaction() {
+        Lock lock = readWriteLock.writeLock();
+        lock.lock();
```

Comment:
It took me a while to understand how the locking here works, but I think I get it, if I may:

1. Reads can return a future. When a read is scheduled, it gets a handle to the lock, adds the future to a "pending reads" lists, and then releases the lock
2. Writes do not return a future. A write grabs the lock, applies the mutation, and then releases the lock.
3. When `syncTransaction` is called, the lock is acquired and all outstanding reads are allowed to finish. This ensures that transaction is now unused and safe to commit. The transaction is committed, then the object's current transaction is replaced by a new one to use in future reads and writes.

I think these semantics make sense, but I wanted to make sure I had that right.

Reply:
Yes. That is all correct. Additionally, it is a read-write lock, so there can be multiple reads in parallel (not just the futures, but also the thing creating the future). But only one write. And that write can decide on its own to sync, based on size / time, independent of explicit requests for it.